source/base/data_out_base.cc

```diff
@@ -8021,9 +8020,14 @@ DataOutInterface<dim, spacedim>::write_xdmf_file(
       const auto &patches = get_patches();
       Assert(patches.size() > 0, DataOutBase::ExcNoPatches());
 
-      for (it = entries.begin(); it != entries.end(); ++it)
+      // We currently don't support writing mixed meshes:
+      for (const auto &patch : patches)
+        Assert(patch.reference_cell == patches[0].reference_cell,
+               ExcNotImplemented());
```

Comment:
I suspect this will generate unused variable warnings in release mode.
```suggestion
#ifdef DEBUG
      // We currently don't support writing mixed meshes:
      for (const auto &patch : patches)
        Assert(patch.reference_cell == patches[0].reference_cell,
               ExcNotImplemented());
#endif
```